aws_ros1_common/test/parameter_reader_test.cpp

```diff
@@ -0,0 +1,92 @@
+    }
+    void TearDown() {
+        producer.deleteParam(PARAM_READER_TEST__PARAM_PREFIX "/" PARAM_READER_TEST__PARAM_KEY);
+        consumer.deleteParam(PARAM_READER_TEST__PARAM_PREFIX "/" PARAM_READER_TEST__PARAM_KEY);
+    }
+    ros::NodeHandle producer;
+    ros::NodeHandle consumer;
+};
+
+TEST_F(ParameterReaderTest, parameterPathResolution)
+{
+    auto parameter_reader = std::make_shared<Aws::Client::Ros1NodeParameterReader>();
+    auto param_path_flat = ParameterPath(
+        "test_parameter_reader/producers/" PARAM_READER_TEST__PARAM_PREFIX "/" PARAM_READER_TEST__PARAM_KEY
+    );
+    auto param_path_variadic = ParameterPath(
+        "test_parameter_reader", "producers", PARAM_READER_TEST__PARAM_PREFIX, PARAM_READER_TEST__PARAM_KEY
+    );
+    auto param_path_complex_no_node_ns = ParameterPath(std::vector<std::string>{},
+            std::vector<std::string>{PARAM_READER_TEST__PARAM_PREFIX, PARAM_READER_TEST__PARAM_KEY});
+    auto param_path_complex_with_node_ns = ParameterPath(
+            std::vector<std::string>{"test_parameter_reader", "producers"},
+            std::vector<std::string>{PARAM_READER_TEST__PARAM_PREFIX, PARAM_READER_TEST__PARAM_KEY});
+
+    ASSERT_EQ(param_path_flat.get_resolved_path('/', '/'), param_path_variadic.get_resolved_path('/', '/'));
+    ASSERT_EQ(param_path_flat.get_resolved_path('/', '/'), param_path_complex_with_node_ns.get_resolved_path('/', '/'));
+
+    std::string flat_result;
+    parameter_reader->ReadStdString(param_path_flat, flat_result);
+    std::string variadic_result;
+    parameter_reader->ReadStdString(param_path_variadic, variadic_result);
+    std::string complex_no_node_ns_result;
+    parameter_reader->ReadStdString(param_path_complex_no_node_ns, complex_no_node_ns_result);
+    std::string complex_with_node_ns_result;
+    parameter_reader->ReadStdString(param_path_complex_with_node_ns, complex_with_node_ns_result);
+
+    ASSERT_EQ(flat_result, variadic_result);
+    ASSERT_EQ(variadic_result, complex_with_node_ns_result);
+    ASSERT_EQ(flat_result, PARAM_READER_TEST__PARAM_VALUE);
+
+    ASSERT_EQ(std::string(), complex_no_node_ns_result);
+    consumer.setParam(PARAM_READER_TEST__PARAM_PREFIX "/" PARAM_READER_TEST__PARAM_KEY, PARAM_READER_TEST__PARAM_VALUE);
+    parameter_reader->ReadStdString(param_path_complex_no_node_ns, complex_no_node_ns_result);
+    ASSERT_EQ(complex_no_node_ns_result, complex_with_node_ns_result);
+}
+
+TEST(ParameterReader, failureTests)
+{
+    auto parameter_reader = std::make_shared<Aws::Client::Ros1NodeParameterReader>();
+    auto inexistent_path = ParameterPath("I don't exist");
+    std::string inexistent_path_result = PARAM_READER_TEST__PARAM_VALUE;
+    /* Querying for a nonexistent parameter should return NOT_FOUND and the out parameter remains unchanged. */
+    ASSERT_EQ(Aws::AwsError::AWS_ERR_NOT_FOUND, parameter_reader->ReadStdString(inexistent_path, inexistent_path_result));
```

Comment:
Nonexistent seems to be an understatement. It is an invalid format for a path. I wonder if we should still be returning `AWS_ERR_NOT_FOUND` in the case of an invalid path.